workflow/controller/operator.go

```diff
@@ -1418,16 +1418,21 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		} else {
 			woc.wf.Status.MarkTaskResultIncomplete(resultName)
 		}
-	}
 
-	if x, ok := pod.Annotations[common.AnnotationKeyOutputs]; ok {
-		woc.log.Warn("workflow uses legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/")
-		if new.Outputs == nil {
-			new.Outputs = &wfv1.Outputs{}
+		if x, ok = pod.Annotations[common.AnnotationKeyOutputs]; ok {
+			if new.Outputs == nil {
+				new.Outputs = &wfv1.Outputs{}
+			}
+			if err := json.Unmarshal([]byte(x), new.Outputs); err != nil {
+				new.Phase = wfv1.NodeError
+				new.Message = err.Error()
+			}
 		}
-		if err := json.Unmarshal([]byte(x), new.Outputs); err != nil {
-			new.Phase = wfv1.NodeError
-			new.Message = err.Error()
+
+		if x, ok = pod.Annotations[common.AnnotationKeyProgress]; ok {
+			if p, ok := wfv1.ParseProgress(x); ok {
+				new.Progress = p
+			}
```

Comment:
can we add a comment above this whole code block (above line 1413) to say something like "if `AnnotationKeyReportOutputsCompleted` is set, it means RBAC prevented `WorkflowTaskResult` from being written: therefore, we use Annotations to get information from Pod"

Reply:
There is already a warning log above.
https://github.com/argoproj/argo-workflows/blob/16f0a8ea3cc0690d32994d309ac6b6e26a9409ca/workflow/controller/operator.go#L1425

Reply:
I see, but I find that comment and the link it references to not be super clear; and moreover, the part I wanted to clarify was that you're using the fact that `pod.Annotations[common.AnnotationKeyReportOutputsCompleted]` is set to mean that we need to use Annotations to get all of the information. I don't find all of this necessarily obvious personally.

Reply:
ok. I've added some comments to clarify this.